text/0000-target-stages.md

```diff
@@ -0,0 +1,146 @@
+flag. This flag would be denoted a special role in the compiler and it would have a special place in the incremental system.
+
+The flag denotes which target the compiler can go up to, and serves as an early indicator of how the build was performed. This moves the burden of stage-indication from the compiler flags, and gives a reliant way for future compilations to evaluate if they should run, load a dependency graph into memory, or start from scratch altogether.
+
+Disconnecting compiler flags from an early recompilation judgement allows us to almost completely overlap the various
+compiler activities of different levels, and allow for more flexibility in activities other than `build` and `check`,
+such as `clippy` or other existing and future projects that may need to partially build with the Rust compiler.
+
+---
+
+While not all cases overlap with "lower" compiler-adjacent activites (`check` would be lower than `build`) and those
+cases would need a recompilation anyway, this allows us to [perform further optimizations][#further-optimizations] than just holding the two workflows. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+The `--target-stage` flag is an option that instructs the compiler to only go so far when compiling your code. This
+allows the compiler to perform numerous optimizations to avoid redundant recompilations.
+
+Users don't usually have to worry about this flag, as it's handled automatically by Cargo via `cargo check` and `cargo build`.
+For an explanation for RFC readers, see [Motivation](#motivation).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are two parts to this feature, one handled in the compiler itself and another handled by the incremental system
+and heavily interacting with Cargo.
+
+
+The easiest part is handled by the compiler, where "safe points to exit" are designated (such as in between of parsing
+to AST and macro-expansion, or lowering THIR to MIR). Depending on the input that the `--target-stage` option takes,
+we exit at one point or another.
+
+Along with the public portion of the RFC, we also have a private new construct, called "Stage dependencies".
+"Stage dependencies" are designated fields that, while they are handled by `Session`, they are not taken into account
+when hashing, and thus, their values are not always loadbearing.
+
+This means that stage dependencies only affect the hash of the necessary stage that depends on them.
+Stage dependencies are also overlapping in the majority of cases, we'll look into it later.
+
+For example, let's say that the user calls `cargo check` two times, with different `-Cstrip` values.
+`cargo check` (in the local module) will not be affected by this change, as `-Cstrip` is a stage dependency on `linking` and beyond.
+Being that `cargo check` doesn't change behaviour depending on the linking behaviour, it should **not** be recompiled.
```

Comment:
Is this talking purely in terms of the incremental compilation cache?  For cargo, we write new cache entries for changes in profile settings, see https://doc.rust-lang.org/nightly/nightly-rustc/cargo/core/compiler/fingerprint/index.html#fingerprints-and-unithashs

Reply:
> Is this talking purely in terms of the incremental compilation cache?

Yes, it's talking from the perspective of rustc's cache specifically.

Is there a reason why fingerprints aren't based purely on the final flags passed to `rustc` (including RUSTFLAGS)? A rustc invocation with two different profiles but with manually set-up flags to be identical should be ~identical for cache purposes.